dynamic_graph_service/src/service/channel/sample_publisher.cc

```diff
@@ -181,56 +128,57 @@ seastar::future<> SamplePublisher::Publish(
 
   for (auto &info : infos) {
     assert(info.worker_id < worker_records_.size());
-    if (!worker_record_id_set_[info.worker_id].count(info.record_id)) {
-      auto dst_pid = partitioner_.GetPartitionId(
-        batch[info.record_id].key.pkey.vid);
-      worker_records_[info.worker_id][dst_pid].push_back(
+    worker_records_[info.worker_id].emplace_back(
         &batch[info.record_id]);
-      worker_record_id_set_[info.worker_id].insert(info.record_id);
-    }
   }
 
   std::vector<uint32_t> pr_ids;
   std::vector<seastar::future<>> futures;
   cppkafka::MessageBuilder builder(kafka_topic_);
 
   for (WorkerId wid = 0; wid < worker_records_.size(); ++wid) {
-    auto& part_records = worker_records_[wid];
-    auto& sink_kafka_partitions = worker_sink_kafka_partitions_[wid];
-    for (PartitionId pid = 0; pid < part_records.size(); ++pid) {
-      if (part_records[pid].empty()) {
-        continue;
+    auto& wr = worker_records_[wid];
+    if (wr.empty()) {
+      continue;
+    }
+
+    // If worker updates size is too large, slice them into small batches.
+    std::vector<actor::BytesBuffer> sliced_batches;
+    size_t slice_size = 0;
+    size_t begin = 0, end = 0;
+    while (end < wr.size()) {
+      slice_size += wr[end]->Size();
+      end++;
+      if (slice_size > batch_size_) {
+        sliced_batches.emplace_back(
+            io::SampleUpdateBatch::Serialize(wr.data() + begin, end - begin));
+        begin = end;
+        slice_size = 0;
       }
-      // Get the sink kafka partition id
-      // TODO(@goldenleaves): double-check
-      auto dst_kafka_pid = sink_kafka_partitions[pid];
+    }
+    if (end > begin) {
+      sliced_batches.emplace_back(
+          io::SampleUpdateBatch::Serialize(wr.data() + begin, end - begin));
```

Comment:
The operations of slicing batches can be integrated into line 129 ~ line 133 ? Then you don't have to iterate all the records twice.

Reply:
two different stages:
  stage 1: collect updates by worker id
  stage 2: slice worker updates by batch
on each stage, the partitioning and indexing logic is one-dimensional, if you want to put stage 2's logic into stage 1, you should maintain a two-dimensional index, the code will be more ugly